### DISCUSSION TITLE
Hooks Outside PiccoloCrud

### DISCUSSION THREAD

**Discussion by Fidel-C** [piccolo-orm/piccolo_api#316] [Ideas]
Just wondering whether Hooks can be emplemented externally in any part of the app. Something like Django signals. So that wherever a record  changes or is added in the database, even if the action was called from the admin panel, the Hook would still be triggered . I think this restriction to PiccoloCrud alone is very limiting. I wish there could be a Django-like approach to this.

**Comment by dantownsend** [comment]
There isn't an official way of doing it at the moment - it's possible to override the save method on a table, but a bit tricky.

Having a proper API for it would be nice, as it's a much more common requirement than I expected.

**Comment by Fidel-C** [comment]
Looking forward to that then.

**Comment by Fidel-C** [comment]
A code example showing how to override the save method on a table will be good. I can't see how to that.

**Comment by dantownsend** [comment]
It's something like this:

```python
from piccolo.utils.sync import run_sync

def save(self, columns=None):
    save_ = super().save

    class Save:
        async def run(self, *args, **kwargs):
            # Add custom logic here

            await save_(columns=columns).run(*args, **kwargs)

        def run_sync(self, *args, **kwargs):
            return run_sync(self.run(*args, **kwargs))

        def __await__(self):
            return self.run().__await__()

    return Save()

```

The reason it looks a bit complicated is because Piccolo lets you run a query as sync or async.

**Comment by Fidel-C** [comment]
Thanks a lot. I observed that  the run method is called when adding records. But doesn't seem to be called when updating existing records.
Quite complicated, just like you said.

**Comment by sinisaos** [comment]
@Fidel-C You can also try something [like this](https://github.com/piccolo-orm/piccolo/issues/987#issuecomment-2124630523). It is based on the @dantownsend code with minor modifications to use (override) also `save` (both `save` and `update`) and `remove` methods.

**Comment by Fidel-C** [comment]
@sinisaos  Thanks a lot.
